tests/virt_autotest/vgpu.pm

```diff
@@ -0,0 +1,368 @@
+        # Create a CI
+        my $ci_id = create_compute_instance($gi_id, $gpu_id);
+    }
+
+    # Create a vGPU device
+    my $vgpu = create_vgpu($gpu_device, get_var('GPU_MIG_MODE'));
+
+    save_original_guest_xmls();
+    my $vgpu_grid_driver = get_required_var("VGPU_GRID_DRIVER_URL");
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        record_info("Guest $guest");
+
+        # Assign vGPU to guest
+        my $gpu_slot_in_guest = "0x0a";
+        assign_vgpu_to_guest($vgpu, $guest, $gpu_slot_in_guest);
+
+        # Install vGPU grid driver in guest
+        install_vgpu_grid_driver($guest, $vgpu_grid_driver);
+
+        # Detach vGPU from guest
+        detach_vgpu_from_guest($gpu_slot_in_guest, $guest);
+        check_guest_health($guest);
+    }
+
+    # Remove vGPU device
+    remove_vgpu($vgpu);
+
+    # Disable GPU SR-IOV
+    assert_script_run("/usr/lib/nvidia/sriov-manage -d $gpu_device");
+    record_info("SRIOV disabled", $gpu_device);
+
+    # Upload vgpu related logs
+    upload_virt_logs($log_dir, "logs");
+
+    # Redefine guest from their original configuration files
+    restore_original_guests();
+}
+
+sub is_host_ready_for_vgpu_test {
+    if (is_sle('<15-SP3') or !is_kvm_host) {
+        record_info("Host not supported!", "NVIDIA vGPU is only supported on KVM hosts with SLE15SP3 and greater!", result => 'softfail');
+        return;
+    }
+    #check VT-d is supported in Intel x86_64 machines
+    if (script_run("grep -m 1 Intel /proc/cpuinfo") == 0) {
+        assert_script_run "dmesg | grep -E \"DMAR:.*IOMMU enabled\"";
+    }
+    script_run("lsmod | grep -e vfio -e nvidia -e mdev");
+    return 'yes';
+}
+
+sub get_gpu_device {
+    #NVIDIA AMPERE GPU
+    my $gpu_devices = script_output("lspci | grep NVIDIA | cut -d ' ' -f1");
+    foreach (split("\n", $gpu_devices)) {
+        return $_ if script_run("lspci -v -s $_ | grep 'SR-IOV'") == 0;
+    }
+}
+
+sub install_vgpu_manager_and_reboot {
+    my ${dirver_url} = shift;
+    die "The vGPU driver URL requires to be in 'http://...' format!" unless ${dirver_url} =~ /http:\/\/.*\/(.*\.run)/;
+    my ${driver_file} = $1;
+
+    # Check if vGPU manager has been already loaded
+    if (script_run("nvidia-smi") == 0) {
+        my ${driver_version} = script_output("nvidia-smi -q | grep 'Driver Version' | grep -oE \"[0-9.]+\"");
+        if (${driver_file} =~ /${driver_version}/) {
+            record_info("Warning", "vGPU manager ${driver_version} has already been loaded!", result => 'softfail');
+            return;
+        }
+    }
+    # Install vGPU manager
+    download_script(${driver_file}, script_url => ${dirver_url});
+    zypper_call "in kernel-default-devel gcc";
+    assert_script_run("modprobe vfio_pci_core") if is_sle('>=15-SP5');
+    assert_script_run("./${driver_file} -s");
+    record_info("vGPU manager is installed successfully.");
+
+    # Reboot host(it took 180+ seconds for ph052 to bootup with calltrace)
```

Comment:
Did you consider using reboot_and_waitup_normal func? Or you prefer simplicity so write a new reboot handling 
 process? It is also fine if stable enough.

Reply:
I thought of that function. I decided to write rebooting statement on my own because `reboot_and_waitup_normal` is too long as it includes the host upgrade tests and perf tests in login_console.pm. In addiiton, my reboot is rather simple, no grub or xen or other things involved.

Reply:
Understand and makes sense. How about putting it to lib for future reuse of simpler reboot handling?

Reply:
I'll update.